### PR TITLE
[WIP] Symmetry with constraints + Started to dvlp dynamics with contacts

### PR DESCRIPTION
**Pull request by paulWegiel**

### CODE REVIEW

File: biorbd_optim/constraints.py [hunk 1]
```diff
@@ -71,11 +71,17 @@ def add_constraints(ocp, nlp):
             elif elem[0] == Constraint.Type.PROPORTIONAL_Q:
                 Constraint.__proportional_variable(ocp, nlp, x, elem[2])
             elif elem[0] == Constraint.Type.PROPORTIONAL_CONTROL:
-                # TODO: Paul = Raise error if INSTANT.END is used
+                if elem[1] == Constraint.Instant.END:
+                    raise RuntimeError(
+                        "A proportional control does not make sense at the last node (Instant.END)"
```
Comment (pariterre) [inline biorbd_optim/constraints.py:76]: That is a bit harsh!

File: biorbd_optim/constraints.py [hunk 2]
```diff
@@ -104,11 +104,26 @@ def __markers_to_pair(ocp, nlp, X, policy):
...
+        :param policy: A tuple or a tuple of tuples whose first two elements are the indexes of elements to be linked proportionally.
+        The third element of each tuple (policy[i][2]) is the proportionality coefficient.
+        """
+        if isinstance(policy[0], tuple):
+            for elem in policy:
+                for v in V:
+                    v = nlp["dof_mapping"].expand(v)
+                    ocp.g = vertcat(ocp.g, v[elem[0]] - elem[2] * v[elem[1]])
+                    ocp.g_bounds.min.append(0)
+                    ocp.g_bounds.max.append(0)
```
Comment (pariterre) [inline biorbd_optim/constraints.py:120]: Do not duplicate the code, put the policy in a list if not tuple and reuse the code

File: biorbd_optim/dynamics.py [hunk 3]
```diff
@@ -3,6 +3,21 @@
...
+        Returns q, qdot, tau (unreduced by a potential symmetry) and qdot_reduced
+        from states, controls and mapping through nlp to condense this code.
+        """
+        nq = nlp["dof_mapping"].nb_reduced
+        q = nlp["dof_mapping"].expand(states[:nq])
+        qdot_reduced = states[nq:]
+        qdot = nlp["dof_mapping"].expand(qdot_reduced)
+        tau = nlp["dof_mapping"].expand(controls[: nlp["nbTau"]])
+
+        return q, qdot, qdot_reduced, tau
```
Comment (pariterre) [inline biorbd_optim/dynamics.py]: Move this to the end of the class

File: biorbd_optim/dynamics.py [hunk 4]
```diff
@@ -19,11 +19,11 @@ def __dispatch_data(states, controls, nlp):
         return q, qdot, qdot_reduced, tau
 
     @staticmethod
-    def forward_dynamics_torque_driven(states, controls, nlp):
+    def forward_dynamics_torque_driven_no_contact(states, controls, nlp):
```
Comment (pariterre) [inline biorbd_optim/dynamics.py]: Do not add "no_contact"

File: biorbd_optim/constraints.py [hunk 5]
```diff
@@ -127,8 +127,28 @@ def __proportional_variable(ocp, nlp, V, policy):
 
     @staticmethod
     def __contact_force_inequality(ocp, nlp, X, U, policy):
-        for i in range(len(U)):
-            contact_forces = Dynamics.get_forces_from_contact(X[i], U[i], nlp)
+        """
+        A compléter, notamment policy est soit un tuple ou un tuple de tuples, avec en 1er indice
+        le numéro de la force de contact et en 2e indice la borne associée
```
Comment (pariterre) [inline biorbd_optim/constraints.py:132]: Even though this is for yourself, write your comment in english. So if you ever forget, this does not lie there forever ;)

File: biorbd_optim/constraints.py [hunk 6]
```diff
@@ -127,8 +127,28 @@ def __proportional_variable(ocp, nlp, V, policy):
...
+                    ocp.g_bounds.min.append(elem[1])
+                    ocp.g_bounds.max.append(10000000)         # Comment ne mettre qu'une borne inf ?
+        else:
+            for i in range(len(U)):
+                contact_forces = Dynamics.get_forces_from_contact(X[i], U[i], nlp)
+                contact_forces = contact_forces[:6] # A changer : il faut réduire par symétrie (ssi sym par construction)
+
+                ocp.g = vertcat(ocp.g, contact_forces[policy[0]])
+                ocp.g_bounds.min.append(policy[1])
+                ocp.g_bounds.max.append(10000000)  # Comment ne mettre qu'une borne inf ?
```
Comment (pariterre) [inline biorbd_optim/constraints.py:151]: Do no duplicate code

File: biorbd_optim/constraints.py (hunk 2 above)
Comment (pariterre) [reply biorbd_optim/constraints.py:120]: Also test isinstance(policy[0], (tuple, list))

File: biorbd_optim/constraints.py (hunk 1 above)
Comment (paulWegiel) [reply biorbd_optim/constraints.py]: Changed in "Instant.END is used even though there is no control u at last node"

File: biorbd_optim/constraints.py (hunk 2 above)
Comment (paulWegiel) [reply biorbd_optim/constraints.py:122]: Code condensed + now tests tuple and lists + raises error if there is a unusable mix of tuples/lists and non tuples/lists in policy.

File: biorbd_optim/dynamics.py (hunk 3 above)
Comment (paulWegiel) [reply biorbd_optim/dynamics.py]: Done

File: biorbd_optim/dynamics.py (hunk 4 above)
Comment (paulWegiel) [reply biorbd_optim/dynamics.py:7]: Changed

File: biorbd_optim/constraints.py (hunk 5 above)
Comment (paulWegiel) [reply biorbd_optim/constraints.py:134]: Done in commit 12959992ef966abe6ceb5969ea1e50577220a7f2

File: biorbd_optim/constraints.py (hunk 2 above)
Comment (paulWegiel) [reply biorbd_optim/constraints.py]: [EDIT] I left a mistake in this, fixed it in the commit 6b3626f97ad5b278a6471610d51d2e2b325b671a

File: biorbd_optim/constraints.py (hunk 6 above)
Comment (pariterre) [reply biorbd_optim/constraints.py]: Could you try math.inf?

**Pull request merged**